doc/reading/tma-modeling-in-depth.md

```diff
@@ -0,0 +1,153 @@
+  In the example in Figure 1, it is $(12, 16)$.
+- $\vec{bs}$ stands for "Box Size". It is a vector of the same dimensionality as $\vec{x}$.
+  In the example in Figure 1, it is $(6, 4)$.
+- $\vec{gr}$ stands for "Global stRide". It is a vector of the same dimensionality as $\vec{x}$.
+  In the example in Figure 1, it is $(1, 14)$.
+- $\vec{er}$ stands for "Element stRide". It is a vector of the same dimensionality as $\vec{x}$.
+  In the example in Figure 1, it is $(1, 1)$ for the left diagram, and $(1, 3)$ for the right diagram.
+
+We call inputs, parameters, and $op$ differently because in the implementation,
+inputs are provided as operands for the PTX instruction,
+parameters are encoded inside the TensorMap descriptor,
+and $op$ defines which PTX instruction to use.
+When looking at the kernel level, only $\vec{x}$ and $sa$ can change;
+parameters and op are predefined constants.
+
+The thing that this $tma$ function does is demonstrated in the CodeBlock 1 below
+(here, assuming the TMA is 2D, we can easily generalize to other dimensionalities):
+
+```python
+def tma(x, sa, *, ga, gs, bs, gr, er, op):
+    ts = [ceildiv(bs[0], er[0]), ceildiv(bs[1], er[1])] # tile size
+    for i0 in range(ts[0]):
+        for i1 in range(ts[1]):
+            smem_idx = i1 * ts[0] + i0
+            gmem_idx0 = x[0] + i0 * er[0]
+            gmem_idx1 = x[1] + i1 * er[1]
+            gmem_idx = gmem_idx0 * gr[0] + gmem_idx1 * gr[1]
+            if op == "load":
+                if gmem_idx0 < gs[0] and gmem_idx1 < gs[1]:
+                    sa[smem_idx] = ga[gmem_idx]
+                else:
+                    sa[smem_idx] = 0
+            else:
+                if gmem_idx0 < gs[0] and gmem_idx1 < gs[1]:
+                    ga[gmem_idx] = sa[smem_idx]
+```
+
+## Predication and correctness
+
+We all know that TMA has built-in boundary check and automatic zero filling for out-of-boundary items.
+But what does this mean? Does it mean that we don't need to predicate a TMA expression at all?
+Does it mean that we don't need to initialize the output buffer of TMA?
+This section discusses these questions.
+
+First to note is, what we should do depend on how we will use the output of TMA.
+The theory for this is called [correctness model](divisibility-of-split.md#allocation-and-correctness-model).
+Specifically, we are interested in:
+
+- For an unpredicated TMA expression, if we do not initialize the buffer, what correctness model can we achieve?
+- In the case where strong correctness is needed, but we have only achieved weak correctness,
+  is there any way to upgrade to strong correctness by doing something?
+
+As we can see from CodeBlock 1, TMA has builtin predicates checking that the indices of all partitioned IterDomains are in bound.
+That is, TMA will never do out-of-boundary access on global memory even if the indices of
+some IterDomains may be out of boundary.
+Therefore, we have:
+
+**Theorem 1:** TMA provides weak correctness.
```

Comment:
Currently, we eliminated all predicates of TMA, and not initializing the output smem buffer. So we are assuming weak correctness. I believe in the future, we need to check the correctness model we need based on the fusion.